### PR TITLE
[REVIEW] Implement string REPLACE

### PR DESCRIPTION
**Pull request by beckernick**
This PR:
- Implements the REPLACE operator for string columns using cudf::strings::replace
- Adds a new end-to-end test in stringsTests.py, and updates the runTest.py
- Refactors the removal of the string encaspulation characters (i.e., the single quotes in  `LIKE '%the%'`) in several parts of LogicalProjection.cpp to use a string utility function

If the implementation looks fine, I'll push the new parquet file to https://github.com/BlazingDB/blazingsql-testing-files and update the CHANGELOG to unblock gpuCI.

This closes https://github.com/BlazingDB/blazingsql/issues/1175

- [x] passes e2e tests locally
- [x] PR to update testing files https://github.com/BlazingDB/blazingsql-testing-files/pull/1
```python
from pyspark.sql import SparkSession
from blazingsql import BlazingContext
import pandas as pd
​
​
# spark = SparkSession.builder \
#     .master("local") \
#     .getOrCreate()
​
# bc = BlazingContext()
​
​
df = pd.DataFrame({
    "a": ["Felipe", "William", "Rodrigo"],
    "b": [2, 4, 6],
    "c": ["2020-11-20", "2020-11-19", "2020-11-18"]
})
​
bc.create_table("df", df)
sdf = spark.createDataFrame(df)
sdf.createOrReplaceTempView("df")
​
query = """
SELECT
    a,
    REPLACE(a, 'i', '##') as a_new,
    c,
    REPLACE(c, '2020', '1999') as c_new
FROM df
"""
​
spark.sql(query).show()
​
print(bc.explain(query))
print(bc.sql(query))
+-------+---------+----------+----------+
|      a|    a_new|         c|     c_new|
+-------+---------+----------+----------+
| Felipe|  Fel##pe|2020-11-20|1999-11-20|
|William|W##ll##am|2020-11-19|1999-11-19|
|Rodrigo| Rodr##go|2020-11-18|1999-11-18|
+-------+---------+----------+----------+

LogicalProject(a=[$0], a_new=[REPLACE($0, 'i', '##')], c=[$1], c_new=[REPLACE($1, '2020', '1999')])
  BindableTableScan(table=[[main, df]], projects=[[0, 2]], aliases=[[a, a_new, c, c_new]])

         a      a_new           c       c_new
0   Felipe    Fel##pe  2020-11-20  1999-11-20
1  William  W##ll##am  2020-11-19  1999-11-19
2  Rodrigo   Rodr##go  2020-11-18  1999-11-18
```

### CODE REVIEW

**Comment by ajschmidt8** [comment]
rerun tests

**Comment by beckernick** [comment]
thanks @ajschmidt8 . iIve intentionally not triggered CI for this PR yet.

If the implementation looks fine, I'll push the new parquet file to https://github.com/BlazingDB/blazingsql-testing-files and update the CHANGELOG to unblock gpuCI.

**Comment by ajschmidt8** [comment]
> thanks @ajschmidt8 . I've intentionally not triggered CI for this PR yet.
> 
> If the implementation looks fine, I'll push the new parquet file to https://github.com/BlazingDB/blazingsql-testing-files and update the CHANGELOG to unblock gpuCI.

whoops! apologies. I'll leave it up to the Blazing guys to make that determination. Looks like there's a CI bug in one of our scripts here that I'll need to address regardless.

**Comment by beckernick** [comment]
Ah, if it's useful for you I'm happy to update the CHANGELOG

**Comment by ajschmidt8** [comment]
> Ah, if it's useful for you I'm happy to update the CHANGELOG

Nope, all good. I've got what I need. Thanks Nick.

File: engine/src/execution_graph/logic_controllers/LogicalProject.cpp
Comment (williamBlazing) [inline engine/src/execution_graph/logic_controllers/LogicalProject.cpp:115]: You should validate that the string is indeed encapsulated by single quotes. I am assuming it usually will, but in case int he future we change something up, we would want to make sure we are not stripping the first and last characters of the string of interest. 
You can also use the string utility `removeEncapsulation` from StringUtils.h

**Review by williamBlazing — CHANGES_REQUESTED** [review]

File: engine/src/execution_graph/logic_controllers/LogicalProject.cpp
Comment (beckernick) [reply engine/src/execution_graph/logic_controllers/LogicalProject.cpp:115]: Thanks for the tip. Should other parts of the codebase be using this utility as well? The pattern `arg_tokens[1].substr(1, arg_tokens[1]` comes up quite a bit in LogicalProject

File: tests/BlazingSQLTest/EndToEndTests/stringTests.py
Comment (williamBlazing) [inline tests/BlazingSQLTest/EndToEndTests/stringTests.py:202]: Please also add some test queries where you are applying a nested transformation and you have REPLACE as the internal transformation and the exeternal transformation 

For example doing something like this:
REPLACE(SUBSTRING(c_comment, 2, 10), 'e', '&&')
SUBSTRING(REPLACE(c_comment, 'e', '&&'), 2, 10)
REPLACE(c_comment,  'a', '1') LIKE '%the%'

**Review by williamBlazing — CHANGES_REQUESTED** [review]

File: engine/src/execution_graph/logic_controllers/LogicalProject.cpp
Comment (williamBlazing) [reply engine/src/execution_graph/logic_controllers/LogicalProject.cpp:115]: Yeah, it would be best to use the utility more in the codebase

**Comment by williamBlazing** [comment]
With those improvements to the PR, then I would say the CHANGELOG can be updated so that we can see the tests run and make sure everything is fine

File: engine/src/execution_graph/logic_controllers/LogicalProject.cpp
Comment (beckernick) [reply engine/src/execution_graph/logic_controllers/LogicalProject.cpp:115]: Sounds good. I'll probably update the rest of this file then

Comment (beckernick) [reply engine/src/execution_graph/logic_controllers/LogicalProject.cpp]: Resolved by https://github.com/BlazingDB/blazingsql/pull/1193/commits/da60b5de0420bb6b6a08cdd7d5aa1bfbcbb70c0e

File: tests/BlazingSQLTest/EndToEndTests/stringTests.py
Comment (beckernick) [reply tests/BlazingSQLTest/EndToEndTests/stringTests.py:202]: Resolved by https://github.com/BlazingDB/blazingsql/pull/1193/commits/952e91232ba46a12b24828ca8be425b37856191c

**Comment by beckernick** [comment]
What's the protocol on adding the new parquet test files to https://github.com/BlazingDB/blazingsql-testing-files ? If I trigger CI before adding them, CI will fail on these new tests. But it feels odd to add them beforehand.

**Comment by beckernick** [comment]
Testing files PR: https://github.com/BlazingDB/blazingsql-testing-files/pull/1

**Comment by felipeblazing** [comment]
rerun tests

**Comment by felipeblazing** [comment]
rerun tests

**Review by felipeblazing — APPROVED** [review]

**Comment by felipeblazing** [comment]
> What's the protocol on adding the new parquet test files to https://github.com/BlazingDB/blazingsql-testing-files ? If I trigger CI before adding them, CI will fail on these new tests. But it feels odd to add them beforehand. Perhaps I'll trigger so that we can see if any other tests fail in CI

addingn them before hand is fine. you can also trigger builds after commiting them by commenting `rerun tests`

**Comment by felipeblazing** [comment]
> > What's the protocol on adding the new parquet test files to https://github.com/BlazingDB/blazingsql-testing-files ? If I trigger CI before adding them, CI will fail on these new tests. But it feels odd to add them beforehand. Perhaps I'll trigger so that we can see if any other tests fail in CI
> 
> addingn them before hand is fine. you can also trigger builds after commiting them by commenting `rerun tests`

and apparently you should never use that expression at all because it triggers the test.

**Comment by beckernick** [comment]
I think quoting a comment that says it will also trigger CI as it's in the texts 😄

**Pull request merged**
